Leaderboards on the hub: Vectara (#1741)
huggingface/blog: leaderboards-on-the-hub-vectara.md

Side note: *On a meaningful side note to us here at Vectara, the HHEM was named after our peer Simon Hughes, who passed away in Nov. of 2023 without notice of natural causes; we decided to name it in his honor due to his [lasting legacy](https://diginomica.com/final-thoughts-memory-simon-mark-hughes-hallucination-research-pioneer) in this space.*

# Setting up HHEM with the LLM leaderboard template
To set up the Vectara HHEM leaderboard, we had to follow a few steps, adjusting the HF leaderboard template code to our needs:
1. After cloning the space repository to our own organization, we created two associated datasets, “requests” and “results”; these datasets maintain the requests submitted by users for new LLMs to evaluate, and the results of such evaluation, respectively.
2. We populated the results dataset with existing results from the initial launch, and updated the “About” and “Citations” sections.

For a simple leaderboard, where evaluations results are pushed by your backend to the results dataset, that’s all you need!

As our evaluation is more complex, we then customized the source code to fit the needs of the HHEM leaderboard - here are the details:

> 1. `leaderboard/src/backend/model_operations.py`: This file contains two primary classes - SummaryGenerator and EvaluationModel.

Reviewer: Perhaps highlight instances of SummaryGenerator and EvaluationModel with backtick to note they are code?
Also the (a) and (b) subsections just show as text after the (1) bullet and do not show as sub-bullets.

Author: Should be good now

---
title: "A guide to setting up your own Hugging Face leaderboard: an end-to-end example with Vectara's hallucination leaderboard"
thumbnail: /blog/assets/leaderboards-on-the-hub/thumbnail.png
authors:
- user: ofermend
  guest: true
- user: minseokbae
  guest: true
- user: clefourrier
---

> # Introduction

Reviewer 2: ```suggestion
# A guide to setting up your own Hugging Face leaderboard: an end-to-end example with Vectara's hallucination leaderboard
```

Hugging Face’s [Open LLM Leaderboard](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard) (originally created by Ed Beeching and Lewis Tunstall, and maintained by Nathan Habib and Clémentine Fourrier) is well known for tracking the performance of open source LLMs, comparing their performance in a variety of tasks, such as [TruthfulQA](https://github.com/sylinrl/TruthfulQA) or [HellaSwag](https://rowanzellers.com/hellaswag/).

This has been of tremendous value to the open-source community, as it provides a way for practitioners to keep track of the best open-source models.

> In late 2023, at Vectara we introduced the [Hughes Hallucination Evaluation Model](https://huggingface.co/vectara/hallucination_evaluation_model) (HHEM), an open-source model for measuring the extent to which an LLM [hallucinates](https://vectara.com/glossary-of-llm-terms/#h-hallucinations-aka-llm-hallucinations) (generates text that is nonsensical or unfaithful to the provided source content). Covering both open source models like Llama2 or Mistral-7B as well as commercial models like OpenAI’s GPT-4, Anthropic Claude, or Google’s Gemini, this model highlighted the stark differences that currently exist between models in terms of their likelihood to hallucinate.

Reviewer 2: ```suggestion
In late 2023, at Vectara we introduced the [Hughes Hallucination Evaluation Model](https://huggingface.co/vectara/hallucination_evaluation_model) (HHEM), an open-source model for measuring the extent to which an LLM [hallucinates](https://vectara.com/glossary-of-llm-terms/#h-hallucinations-aka-llm-hallucinations) (generates text that is nonsensical or unfaithful to the provided source content). Covering both open source models like Llama 2 or Mistral 7B, as well as commercial models like OpenAI’s GPT-4, Anthropic Claude, or Google’s Gemini, this model highlighted the stark differences that currently exist between models in terms of their likelihood to hallucinate.
```

Maybe add links to the Hub for Llama 2 and Mistral, or to the project pages for the others.

As we continue to add new models to HHEM, we were looking for an open-source solution to manage and update the HHEM leaderboard. 

Quite recently, the Hugging Face leaderboard team released leaderboard templates ([here](https://huggingface.co/demo-leaderboard) and [here](https://huggingface.co/demo-leaderboard-backend)). These are lightweight versions of the Open LLM Leaderboard itself, which are both open-source and simpler to use than the original code. 

> Today we’re happy to announce the release of the [new HHEM leaderboard](https://huggingface.co/spaces/vectara/leaderboard), powered by the [HF leaderboard template](https://huggingface.co/demo-leaderboard-backend).

Reviewer 2: Still not public, is it?

Author: cc @ofermend - when do you want to pass it to public?

Reviewer: public now!

> # Vectara’s Hughes Hallucination Evaluation Model (HHEM)

Reviewer 2: ```suggestion
## Vectara’s Hughes Hallucination Evaluation Model (HHEM)
```

> The Hughes Hallucination Evaluation Model (HHEM) Leaderboard is dedicated to assessing the frequency of hallucinations in document summaries generated by Large Language Models (LLMs) such as GPT-4, Google Gemini or Meta’s Llama2. To use it you can follow the instructions [here](https://huggingface.co/vectara/hallucination_evaluation_model).

Reviewer 2: ```suggestion
The Hughes Hallucination Evaluation Model (HHEM) Leaderboard is dedicated to assessing the frequency of hallucinations in document summaries generated by Large Language Models (LLMs) such as GPT-4, Google Gemini or Meta’s Llama 2. To use it you can follow the instructions [here](https://huggingface.co/vectara/hallucination_evaluation_model).
```

> By doing an open-source release of this model, we (at [Vectara](https://vectara.com)) aim to democratize the evaluation of LLM hallucinations, driving awareness to the differences that exist in LLM performance in terms of propensity to hallucinate. 

Reviewer 2: ```suggestion
By doing an open-source release of this model, we at [Vectara](https://vectara.com) aim to democratize the evaluation of LLM hallucinations, driving awareness to the differences that exist in LLM performance in terms of propensity to hallucinate. 
```

> Our initial release of HHEM was a [Huggingface model](https://huggingface.co/vectara/hallucination_evaluation_model) alongside the [Github repository](https://github.com/vectara/hallucination-leaderboard) (which is still maintained), but we quickly realized that we needed a mechanism to allow new types of models to be evaluated. Using the HF leaderboard code template, we were able to quickly put together the new leaderboard that allows for dynamic updates, and we encourage the LLM community to submit new relevant models for HHEM evaluation.

Reviewer 2: ```suggestion
Our initial release of HHEM was a [Huggingface model](https://huggingface.co/vectara/hallucination_evaluation_model) alongside a [Github repository](https://github.com/vectara/hallucination-leaderboard), but we quickly realized that we needed a mechanism to allow new types of models to be evaluated. Using the HF leaderboard code template, we were able to quickly put together a new leaderboard that allows for dynamic updates, and we encourage the LLM community to submit new relevant models for HHEM evaluation.
```

> Side note: *On a meaningful side note to us here at Vectara, the HHEM was named after our peer Simon Hughes, who passed away in Nov. of 2023 without notice of natural causes; we decided to name it in his honor due to his [lasting legacy](https://diginomica.com/final-thoughts-memory-simon-mark-hughes-hallucination-research-pioneer) in this space.*

Reviewer 2: ```suggestion
<br>
<div style="background-color: #e6f9e6; padding: 16px 32px; outline: 2px solid; border-radius: 10px;">
On a meaningful side note to us here at Vectara, the HHEM was named after our peer Simon Hughes, who passed away in Nov. of 2023 without notice of natural causes; we decided to name it in his honor due to his [lasting legacy](https://diginomica.com/final-thoughts-memory-simon-mark-hughes-hallucination-research-pioneer) in this space.
</div>
```

Just a suggestion if you want it to show inside a box like here: https://huggingface.co/blog/introducing-private-hub

Author: Looks nice

> # Setting up HHEM with the LLM leaderboard template

Reviewer 2: ```suggestion
## Setting up HHEM with the LLM leaderboard template
```

To set up the Vectara HHEM leaderboard, we had to follow a few steps, adjusting the HF leaderboard template code to our needs:

> 1. After cloning the space repository to our own organization, we created two associated datasets, “requests” and “results”; these datasets maintain the requests submitted by users for new LLMs to evaluate, and the results of such evaluation, respectively.

Reviewer 2: ```suggestion
1. After cloning the space repository to our own organization, we created two associated datasets: “requests” and “results”; these datasets maintain the requests submitted by users for new LLMs to evaluate, and the results of such evaluations, respectively.
```

I'd maybe link to these datasets (when they are up) so users interested can explore what they look like.

2. We populated the results dataset with existing results from the initial launch, and updated the “About” and “Citations” sections.

For a simple leaderboard, where evaluations results are pushed by your backend to the results dataset, that’s all you need!

As our evaluation is more complex, we then customized the source code to fit the needs of the HHEM leaderboard - here are the details:
1. `leaderboard/src/backend/model_operations.py`: This file contains two primary classes - `SummaryGenerator` and `EvaluationModel`.

>     a. The `SummaryGenerator` generates summaries based on your evaluation dataset and calculates metrics like Answer Rate and Average Summary Length.

Reviewer 2: Is the evaluation dataset a HF dataset?

Author: cc @ofermend ?

Author: It's a private proprietary dataset, I'll edit the blog

    b. The `EvaluationModel` loads our proprietary Hughes Hallucination Evaluation Model (HHEM) to assess these summaries, yielding metrics such as Factual Consistency Rate and Hallucination Rate.
2. `leaderboard/src/backend/evaluate_model.py`: defines the `Evaluator` class which utilizes both `SummaryGenerator` and `EvaluationModel` to compute and return results in JSON format. 

> 3. `leaderboard/src/backend/run_eval_suite.py`: contains a function run_evaluation that leverages `Evaluator` to obtain and upload evaluation results to Vectara (in the “results” dataset mentioned above).

Reviewer 2: ```suggestion
3. `leaderboard/src/backend/run_eval_suite.py`: contains a function `run_evaluation` that leverages `Evaluator` to obtain and upload evaluation results to the `results` dataset mentioned above, causing them to appear in the leaderboard.
```

4. `leaderboard/main_backend.py`: Manages pending evaluation requests and executes auto evaluations using aforementioned classes and functions. It also includes an option for users to replicate our evaluation results.

> The final source code is available in our [HHEM leaderboard repository](https://huggingface.co/spaces/vectara/leaderboard).

Reviewer 2: ```suggestion
The final source code is available in [the Files tab](https://huggingface.co/spaces/vectara/leaderboard/tree/main) of our [HHEM leaderboard repository](https://huggingface.co/spaces/vectara/leaderboard).
```

(and/or link the individual files above)

> With all these changes, we now have the evaluation pipeline ready to go, and easily deployable as a Huggingface space.

Reviewer 2: ```suggestion
With all these changes, we now have the evaluation pipeline ready to go, and easily deployable as a Huggingface Space.
```

> # Summary

Reviewer 2: ```suggestion
## Summary
```
